osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs

```diff
@@ -1,19 +1,528 @@
+            // Remove beats that are too close to the next one (e.g. due to timing point changes)
+            for (var i = beats.Count - 2; i >= 0; i--)
+            {
+                var beat = beats[i];
+
+                if (Precision.AlmostBigger(beatmap.ControlPointInfo.TimingPointAt(beat).BeatLength / 2, beats[i + 1] - beat)) beats.RemoveAt(i);
+            }
+
+            return beats;
+        }
+
+        private void addHitSamples(IEnumerable<OsuHitObject> hitObjects, IReadOnlyList<OsuHitObject> origHitObjects)
+        {
+            var lastSampleIdx = 0;
+
+            foreach (var x in hitObjects)
+            {
+                var samples = getSamplesAtTime(origHitObjects, x.StartTime);
+
+                if (samples == null)
+                {
+                    while (lastSampleIdx < origHitObjects.Count && origHitObjects[lastSampleIdx].StartTime <= x.StartTime)
+                        lastSampleIdx++;
+                    lastSampleIdx--;
+
+                    if (lastSampleIdx < 0 && lastSampleIdx >= origHitObjects.Count) continue;
+
+                    if (lastSampleIdx < origHitObjects.Count - 1)
+                    {
+                        // get samples from the next hit object if it is closer in time
+                        if (origHitObjects[lastSampleIdx + 1].StartTime - x.StartTime < x.StartTime - origHitObjects[lastSampleIdx].StartTime)
+                            lastSampleIdx++;
+                    }
+
+                    x.Samples = origHitObjects[lastSampleIdx].Samples.Where(s => !HitSampleInfo.AllAdditions.Contains(s.Name)).ToList();
+                }
+                else
+                    x.Samples = samples;
+            }
+        }
+
+        private void fixComboInfo(List<OsuHitObject> hitObjects, List<OsuHitObject> origHitObjects)
+        {
+            // First follow the combo indices in the original beatmap
+            hitObjects.ForEach(x =>
+            {
+                var origObj = origHitObjects.FindLast(y => Precision.AlmostBigger(x.StartTime, y.StartTime));
+                x.ComboIndex = origObj?.ComboIndex ?? 0;
+            });
+            // Then reprocess them to ensure continuity in the combo indices and add indices in current combo
+            var combos = hitObjects.GroupBy(x => x.ComboIndex).ToList();
+
+            for (var i = 0; i < combos.Count; i++)
+            {
+                var group = combos[i].ToList();
+                group.First().NewCombo = true;
+                group.Last().LastInCombo = true;
+
+                for (var j = 0; j < group.Count; j++)
+                {
+                    var x = group[j];
+                    x.ComboIndex = i;
+                    x.IndexInCurrentCombo = j;
+                }
+            }
+        }
+
+        private void randomizeCirclePos(IReadOnlyList<OsuHitObject> hitObjects)
+        {
+            if (hitObjects.Count == 0) return;
+
+            var rng = new Random(Seed.Value.GetValueOrDefault());
```

Comment:
I'm pretty sure this isn't doing what is intended. If `Seed.Value` is null, then calling `.GetValueOrDefault()` returns 0. So this is not in fact using some random seed, it's using exactly seed 0 every time.

`new Random(0)` is not equivalent to `new Random()`. The latter [uses tick count for initialisation](https://referencesource.microsoft.com/#mscorlib/system/random.cs,52).

Reply:
`Seed.Value` can never be null at this point because it has already been replaced by a random value if the user doesn't specify a seed. I should move this to a better place to avoid the confusion.

Reply:
There should be an assertion that the seed is not null in that case, instead of falling back to a default that is never supposed to actually happen